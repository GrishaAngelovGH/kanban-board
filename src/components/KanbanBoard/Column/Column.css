.kanban-column {
  min-height: 240px;
  max-height: 600px;
  margin: 25px 10px;
}

.dropdown > .dropdown-toggle::after {
  display: none;
}

.solid-column {
  background: #e3e3e3;
  box-shadow: 0 8px 16px rgba(0, 0, 0, 0.15);
}

.blurred-column {
  backdrop-filter: blur(100px);
  box-shadow: 0px 0px 20px gray;
}

.column-description {
  text-align: justify;
}