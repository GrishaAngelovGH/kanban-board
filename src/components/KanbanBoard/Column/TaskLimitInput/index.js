import TaskLimitInput from "./TaskLimitInput"

export default TaskLimitInput