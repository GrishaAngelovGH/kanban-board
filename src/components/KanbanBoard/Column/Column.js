import { useDroppable } from "@dnd-kit/core"

import Button from "react-bootstrap/Button"
import ButtonGroup from "react-bootstrap/ButtonGroup"
import Dropdown from "react-bootstrap/Dropdown"
import DropdownButton from "react-bootstrap/DropdownButton"

import Task from "./Task"

import settingsRepository from "persistent/persistentSettingsRepository"

import "./Column.css"

const Column = ({
  id, title, description, tasks, markedAsDone,
  onDeleteColumn, onAddTask, onEditTask, onDeleteTask,
  onDeleteAllTasks, onAssignUser, onMarkColumnAsDone
}) => {
  const { isOver, setNodeRef } = useDroppable({ id })

  const handleDeleteAction = () => {
    onDeleteColumn(id)
  }

  const handleAddTask = () => {
    onAddTask(id)
  }

  const handleMarkAsDone = () => {
    onMarkColumnAsDone(id)
  }

  const handleDeleteAllTasks = () => {
    onDeleteAllTasks(id)
  }

  const gridViewClasses = "col-md-5 col-lg-3"
  const singleColumnViewClasses = "col-md-12 col-lg-7"
  const isGridView = settingsRepository.isGridView()
  const hasNoBackground = settingsRepository.hasNoBackground()
  const hasSolidColumnStyle = settingsRepository.hasSolidColumnStyle()
  const layoutClasses = isGridView ? gridViewClasses : singleColumnViewClasses
  const columnStyle = hasSolidColumnStyle ? "solid-column" : "blurred-column"

  // Generally, when "solid" column is chosen, the text should be dark.
  // When the column is set to "blurred" and the background is set to "No Background", the text should be dark.
  // When the column is set to "blurred" and the background is different than "No Background", the text should be light.

  const titleClass = hasSolidColumnStyle || hasNoBackground ? "text-dark" : "text-white"
  const descriptionClass = hasSolidColumnStyle || hasNoBackground ? "text-secondary" : "text-white"
  const checkButtonClass = hasSolidColumnStyle ? "secondary-subtle" : "light"
  const taskLengthButtonClass = hasSolidColumnStyle ? "secondary" : "light"

  return (
    <div
      ref={setNodeRef}
      className={`${layoutClasses} ${columnStyle} rounded p-4 overflow-auto kanban-column`}
      style={{ borderStyle: isOver ? "dashed" : "none" }}
    >
      <div className="row">
        <div className={`${isGridView ? "col-7 col-lg-8" : "col-10"}`}>
          <h3 className={`${titleClass} text-break`}>{title}</h3>
        </div>
        <div className={`${isGridView ? "col-5 col-lg-4" : "col-2"}`}>
          <ButtonGroup size="sm">
            {markedAsDone && <Button variant={checkButtonClass} className="bi bi-check-circle-fill text-success" disabled />}

            <Button variant={taskLengthButtonClass} disabled>{tasks.length}</Button>

            <DropdownButton as={ButtonGroup} size="sm" variant="light" title={<i className="bi bi-three-dots-vertical"></i>}>
              <Dropdown.Item onClick={handleAddTask}>
                <i className="bi bi-plus-circle-fill text-primary mx-1"></i>
                Add Task
              </Dropdown.Item>
              <Dropdown.Item onClick={handleMarkAsDone}>
                <i className="bi bi-check-square-fill text-success mx-1"></i>
                {markedAsDone ? "Unmark as Done" : "Mark as Done"}
              </Dropdown.Item>
              <Dropdown.Item onClick={handleDeleteAction}>
                <i className="bi bi-x-circle-fill text-danger mx-1"></i>
                Delete Column
              </Dropdown.Item>
              <Dropdown.Item onClick={handleDeleteAllTasks}>
                <i className="bi bi-x-circle-fill text-danger mx-1"></i>
                Delete All Tasks
              </Dropdown.Item>
            </DropdownButton>
          </ButtonGroup>
        </div>
      </div>

      <p className={descriptionClass}>{description}</p>

      <div className="overflow-hidden">
        {
          tasks.map(v => (
            <Task
              key={v.id}
              id={v.id}
              columnId={id}
              assignedIds={v.assignedIds}
              title={v.title}
              description={v.description}
              priority={v.priority}
              isGridView={isGridView}
              markedAsDone={markedAsDone}
              onEdit={onEditTask}
              onDelete={onDeleteTask}
              onAssignUser={onAssignUser}
            />
          ))
        }
      </div>
    </div>
  )
}
export default Column