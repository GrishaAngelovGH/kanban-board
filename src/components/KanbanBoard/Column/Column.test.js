import { render } from "@testing-library/react"

import {
  BrowserRouter as Router,
  Routes,
  Route
} from "react-router-dom"

import Column from "./Column"


jest.mock('persistent/persistentUserRepository')

test("should render Column component", () => {
  const view = render(
    <Router>
      <Routes>
        <Route
          path="/"
          element={
            <Column
              title="Column Title"
              description="Column Description"
              limit={0}
              tasks={[
                { id: 1, title: "Task 1", description: "Description 1", assignedIds: [], priority: "medium", isLocked: false },
                { id: 2, title: "Task 2", description: "Description 2", assignedIds: [], priority: "medium", isLocked: false }
              ]}
              columnHandlers={{
                onDeleteColumn: () => { },
                onAddTask: () => { },
                onEditTask: () => { },
                onDeleteTask: () => { },
                onDeleteAllTasks: () => { },
                onAssignUser: () => { },
                onToggleTaskLock: () => { },
                onMarkColumnAsDone: () => { },
                onSetColumnLimit: () => { },
                onColumnUpdate: () => { }
              }}
            />
          }
        />
      </Routes>
    </Router>
  )

  expect(view).toMatchSnapshot()
})