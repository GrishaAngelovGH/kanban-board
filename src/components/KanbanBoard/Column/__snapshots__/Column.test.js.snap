// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Column component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="col-md-5 col-lg-3 blurred-column rounded p-4 overflow-auto kanban-column"
        style="border-style: dashed;"
      >
        <div
          class="row"
        >
          <div
            class="col-7 col-lg-8"
          >
            <h3
              class="text-white text-break"
            >
              Column Title
            </h3>
          </div>
          <div
            class="col-5 col-lg-4"
          >
            <div
              class="btn-group btn-group-sm"
              role="group"
            >
              <button
                class="btn btn-light"
                disabled=""
                type="button"
              >
                2
              </button>
              <div
                class="dropdown btn-group"
                role="group"
              >
                <button
                  aria-expanded="false"
                  class="dropdown-toggle btn btn-light btn-sm"
                  type="button"
                >
                  <i
                    class="bi bi-three-dots-vertical"
                  />
                </button>
              </div>
            </div>
          </div>
        </div>
        <p
          class="text-white"
        >
          Column Description
        </p>
        <div
          class="overflow-hidden"
        >
          <div
            class="row m-0 bg-white border border-3 border-success mt-3 rounded shadow p-1"
          >
            <div
              class="col-9"
            >
              <p
                class="fw-bold"
              >
                Task 1
              </p>
            </div>
            <div
              class="col-3"
            >
              <i
                aria-describedby=""
                aria-disabled="false"
                aria-roledescription="draggable"
                class="bi bi-grip-horizontal fs-2"
                role="button"
                tabindex="0"
              />
            </div>
            <p
              class="text-secondary"
            >
              Description 1
            </p>
            <div
              class="mb-3"
            />
            <div
              class="d-flex justify-content-between"
            >
              <div>
                <i
                  class="bi bi-trash fs-4 text-danger"
                  role="button"
                />
                <i
                  class="bi bi-pencil-square fs-4 text-secondary mx-1"
                  role="button"
                />
                <i
                  class="bi bi-person-circle fs-4 text-secondary mx-1"
                  role="button"
                />
              </div>
              <i
                class="bi bi-bookmark-fill fs-4 text-success"
              />
            </div>
          </div>
          <div
            class="row m-0 bg-white border border-3 border-success mt-3 rounded shadow p-1"
          >
            <div
              class="col-9"
            >
              <p
                class="fw-bold"
              >
                Task 2
              </p>
            </div>
            <div
              class="col-3"
            >
              <i
                aria-describedby=""
                aria-disabled="false"
                aria-roledescription="draggable"
                class="bi bi-grip-horizontal fs-2"
                role="button"
                tabindex="0"
              />
            </div>
            <p
              class="text-secondary"
            >
              Description 2
            </p>
            <div
              class="mb-3"
            />
            <div
              class="d-flex justify-content-between"
            >
              <div>
                <i
                  class="bi bi-trash fs-4 text-danger"
                  role="button"
                />
                <i
                  class="bi bi-pencil-square fs-4 text-secondary mx-1"
                  role="button"
                />
                <i
                  class="bi bi-person-circle fs-4 text-secondary mx-1"
                  role="button"
                />
              </div>
              <i
                class="bi bi-bookmark-fill fs-4 text-success"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="col-md-5 col-lg-3 blurred-column rounded p-4 overflow-auto kanban-column"
      style="border-style: dashed;"
    >
      <div
        class="row"
      >
        <div
          class="col-7 col-lg-8"
        >
          <h3
            class="text-white text-break"
          >
            Column Title
          </h3>
        </div>
        <div
          class="col-5 col-lg-4"
        >
          <div
            class="btn-group btn-group-sm"
            role="group"
          >
            <button
              class="btn btn-light"
              disabled=""
              type="button"
            >
              2
            </button>
            <div
              class="dropdown btn-group"
              role="group"
            >
              <button
                aria-expanded="false"
                class="dropdown-toggle btn btn-light btn-sm"
                type="button"
              >
                <i
                  class="bi bi-three-dots-vertical"
                />
              </button>
            </div>
          </div>
        </div>
      </div>
      <p
        class="text-white"
      >
        Column Description
      </p>
      <div
        class="overflow-hidden"
      >
        <div
          class="row m-0 bg-white border border-3 border-success mt-3 rounded shadow p-1"
        >
          <div
            class="col-9"
          >
            <p
              class="fw-bold"
            >
              Task 1
            </p>
          </div>
          <div
            class="col-3"
          >
            <i
              aria-describedby=""
              aria-disabled="false"
              aria-roledescription="draggable"
              class="bi bi-grip-horizontal fs-2"
              role="button"
              tabindex="0"
            />
          </div>
          <p
            class="text-secondary"
          >
            Description 1
          </p>
          <div
            class="mb-3"
          />
          <div
            class="d-flex justify-content-between"
          >
            <div>
              <i
                class="bi bi-trash fs-4 text-danger"
                role="button"
              />
              <i
                class="bi bi-pencil-square fs-4 text-secondary mx-1"
                role="button"
              />
              <i
                class="bi bi-person-circle fs-4 text-secondary mx-1"
                role="button"
              />
            </div>
            <i
              class="bi bi-bookmark-fill fs-4 text-success"
            />
          </div>
        </div>
        <div
          class="row m-0 bg-white border border-3 border-success mt-3 rounded shadow p-1"
        >
          <div
            class="col-9"
          >
            <p
              class="fw-bold"
            >
              Task 2
            </p>
          </div>
          <div
            class="col-3"
          >
            <i
              aria-describedby=""
              aria-disabled="false"
              aria-roledescription="draggable"
              class="bi bi-grip-horizontal fs-2"
              role="button"
              tabindex="0"
            />
          </div>
          <p
            class="text-secondary"
          >
            Description 2
          </p>
          <div
            class="mb-3"
          />
          <div
            class="d-flex justify-content-between"
          >
            <div>
              <i
                class="bi bi-trash fs-4 text-danger"
                role="button"
              />
              <i
                class="bi bi-pencil-square fs-4 text-secondary mx-1"
                role="button"
              />
              <i
                class="bi bi-person-circle fs-4 text-secondary mx-1"
                role="button"
              />
            </div>
            <i
              class="bi bi-bookmark-fill fs-4 text-success"
            />
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
